### PR TITLE
[0.20] [Path] Add simple test framework for Slot op

### PR DESCRIPTION
**Pull request by Russ4262**
Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

**Comment by Russ4262** [comment]
I am not sure why gcc 7 Travis Slot op test failed and the other two succeeded.  Rebased to rerun the tests.

File: src/Mod/Path/PathTests/TestPathSlot.py [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+import PathScripts.PathJob as PathJob
+import PathScripts.PathSlot as PathSlot
+from PathTests.PathTestUtils import PathTestBase
+
+
+class TestPathSlot(PathTestBase):
+    """Unit tests for the Slot operation."""
+
+    @staticmethod
+    def _setup_test_environment():
```
Comment (mlampert) [inline src/Mod/Path/PathTests/TestPathSlot.py:34]: please don't use `_` names for classes or functions, camel case is the way to go

File: src/Mod/Path/PathTests/TestPathSlot.py [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        slot.LayerMode = "Single-pass"
+
+        slot.recompute()
+        self.doc.recompute()
+        
+        pnts = list()
+        for c in slot.Path.Commands:
+            p = c.Parameters
+            keys = str(p.keys())[9:]
+            if "'X'" in keys and "'Y'" in keys:
```
Comment (mlampert) [inline src/Mod/Path/PathTests/TestPathSlot.py:150]: why not just use `c.Parameters.get('X') and c.Parameter.get('Y')` ?

File: src/Mod/Path/PathTests/TestPathSlot.py [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        if len(pnts) != 2:
+            for c in slot.Path.Commands:
+                p = c.Parameters
+                print("Parameters: {}".format(p))
+            self.assertEqual(len(pnts), 2)
+
+        # Verify each line-segment point, excluding arcs
+        verify_points = ['(7.5, 0.0)', '(7.5, 15.0)']
+        for i in range(2):
+            self.assertEqual(pnts[i], verify_points[i])
```
Comment (mlampert) [inline src/Mod/Path/PathTests/TestPathSlot.py:165]: I highly recommend not comparing doubles directly but use PathGeom to make the test more robust.

File: src/Mod/Path/PathTests/TestPathSlot.py [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    @classmethod
+    def setUpClass(cls):
+        """setUpClass()...
+        This method is called upon instantiation of this test class.  Add code and objects here
+        that are needed for the duration of the test() methods in this class.
+        This method does not have access to the class `self` reference.  This method
+        is able to call static methods within this same class.
+        """
+        # FreeCAD.Console.PrintMessage("TestPathSlot.setUpClass()\n")
+        cls._setup_test_environment()
```
Comment (mlampert) [inline src/Mod/Path/PathTests/TestPathSlot.py:73]: why this layer of indirection?

File: src/Mod/Path/PathTests/TestPathSlot.py [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+        box1.Placement = FreeCAD.Placement(FreeCAD.Vector(5.0, 0.0, 5.0), FreeCAD.Rotation(FreeCAD.Vector(0,0,1), 0))
+        doc.recompute()
+        cut = doc.addObject('Part::Cut', 'Cut')
+        cut.Base = box0
+        cut.Tool = box1
+        doc.recompute()
+
+        # Create Job object
+        PathJob.Create('Job', [doc.Cut], None)
+        doc.recompute()
```
Comment (mlampert) [inline src/Mod/Path/PathTests/TestPathSlot.py:62]: It would be simpler to just add a test file and load that. Would also make the tests simpler since you don't have to search for your feature each time.

File: src/Mod/Path/PathTests/TestPathSlot.py [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+    def _format_point(self, cmd):
+        """Accepts command dict and returns point string coordinate"""
+        x = round(cmd["X"], 2)
+        y = round(cmd["Y"], 2)
+        return "({}, {})".format(x, y)
+
+    def test00(self):
+        '''Test Slot on horizontal rectangular face: ReverseDirection=True; LayerMode=Single-pass'''
+
+        # Identify base feature(s) to be used for operation's Base Geometry
```
Comment (mlampert) [inline src/Mod/Path/PathTests/TestPathSlot.py:123]: the important question is _which_ feature you're looking for ;)

File: src/Mod/Path/PathTests/TestPathSlot.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        # default values used
+
+    def _format_point(self, cmd):
+        """Accepts command dict and returns point string coordinate"""
+        x = round(cmd["X"], 2)
+        y = round(cmd["Y"], 2)
+        return "({}, {})".format(x, y)
+
+    def test00(self):
+        '''Test Slot on horizontal rectangular face: ReverseDirection=True; LayerMode=Single-pass'''
```
Comment (mlampert) [inline src/Mod/Path/PathTests/TestPathSlot.py:121]: It is good practice to include the expected outcome of a test - the description should tell me what the test intends to verify.

File: src/Mod/Path/PathTests/TestPathSlot.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        # Set Operation
+        slot.ReverseDirection = True
+        slot.LayerMode = "Single-pass"
+
+        slot.recompute()
+        self.doc.recompute()
+        
+        pnts = list()
+        for c in slot.Path.Commands:
+            p = c.Parameters
```
Comment (mlampert) [inline src/Mod/Path/PathTests/TestPathSlot.py:148]: what about the `Command` ?
It seems unnecessarily roundabout testing - you know the start point and the end point, so there should be a
* `G0` to safe height
* `G0` to starting point
* `G1` in z down to operation depth
* `G1` to end point
* `G1` (or `G0`) up to safe height
There should be no `G2` or `G3` commands and every other command can be ignored, like comments. As a reader I would immediately see what is being tested.

**Review by mlampert — CHANGES_REQUESTED** [review]

File: src/Mod/Path/PathTests/TestPathSlot.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+        '''
+
+        # Identify base feature to be used for operation's Base Geometry
+        base = self.doc.Cut
+        edge_list = list()
+        for i in range(0, len(base.Shape.Edges)):
+            edge = "Edge%d" % (i+1)
+            e = base.Shape.getElement(edge)
+            eBB = e.BoundBox
+            if (e.Length == 15.0 and round(eBB.ZLength, 6) == 0.0 and round(eBB.ZMin, 6) == 5.0):
```
[the remark this answers is not in the document]
Comment (Russ4262) [reply src/Mod/Path/PathTests/TestPathSlot.py:177]: Are you suggesting I convert the current `if...` statement to `PathGeom.isRoughly(e.Length, 15.0) and PathGeom.isRoughly(eBB.ZLength, 0.0) and PathGeom.isRoughly(eBB.ZMin, 5.0)` ?

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 2 above)
Comment (Russ4262) [reply src/Mod/Path/PathTests/TestPathSlot.py:150]: I want to only work with commands that contain both `X` and `Y` parameters.  I am doing some code scenario testing to see if I understand what you are suggesting here.  I will try to respond with my findings.

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 3 above)
Comment (Russ4262) [reply src/Mod/Path/PathTests/TestPathSlot.py:165]: In this instance, the `self.assertEqual()` is comparing strings as I have seen done elsewhere in Path unit tests.  Should I compare strings differently?

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 4 above)
Comment (Russ4262) [reply src/Mod/Path/PathTests/TestPathSlot.py:73]: I will remove this and combine the methods and add additional comments to explain the purpose of the code.  The one-liner seen here is a nice clear indication of what is being done - that was my thinking, as well as educating those who follow after and want to learn from this unit test.

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 5 above)
Comment (Russ4262) [reply src/Mod/Path/PathTests/TestPathSlot.py:62]: I have not found a good answer for this one due to the challenges your suggestion introduces.  Yes, I do have a test file started and would like to use it to simplify things a bit.  The problem I run into is that when I modify the geometry of the test file I am working on, the edge and face numbering are disrupted, which in turn forces me to have to resynchronize many of the edge and face names used in `test##()` methods.  
. 
So, for now, I am preferring to search for each feature until I have a test file that will not need much maintenance or will not re-enumerate the edge and face names with future additions.

For example:
Snip macro screenshot-149768
This test file is coming along nicely, but when I make additions, the face and edge numbers get disrupted, causing the need to go back into the test module and edit edge and face name references.

My goal is to use a test file.  For now, I would like to get a basic unit test created and functional and merged so we have a starting point and some help with testing changes to the operation.

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 7 above)
Comment (Russ4262) [reply src/Mod/Path/PathTests/TestPathSlot.py:121]: Okay.  Thanks.  I will edit the `test##()` method docStrings throughout all the unit tests I am creating to better communicate the purpose of the test and desired outcome.

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 8 above)
Comment (Russ4262) [reply src/Mod/Path/PathTests/TestPathSlot.py:148]: I don't think I understand what change you are suggesting.  
.
My current strategy is to search all the commands for the two points that should exist within the command set.  
.
Having revisited this Slot op now that I found the issue with the Linux gcc-7 build, I know I need to include depth verification in all the unit tests I am creating.  I will have to figure out how to incorporate such verification across the board in all of them.  I think I have an idea how to do it.

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (mlampert) [reply src/Mod/Path/PathTests/TestPathSlot.py:190]: that would be my recommendation, yes.

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 2 above)
Comment (mlampert) [reply src/Mod/Path/PathTests/TestPathSlot.py]: you're converting all parameters into a string and then check for 'X' and 'Y'.. What I meant is that `Parameters` already is a dictionary, so you can checkk if the keys exist directly.

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 3 above)
Comment (mlampert) [reply src/Mod/Path/PathTests/TestPathSlot.py:178]: you're converting a float to a string - could you end up with something like `'(7.499999999999, 0.00000000001)'` ?

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 5 above)
Comment (mlampert) [reply src/Mod/Path/PathTests/TestPathSlot.py:68]: wow - this is not a  test file, it's a torture file ;)

maybe use multiple files, it's easier to verify multiple simple shapes than one big one which includes everything. sliptonic has a new internet connection, so we're allowed to use more test files ;)

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 2 above)
Comment (Russ4262) [reply src/Mod/Path/PathTests/TestPathSlot.py]: Okay.  Thanks.  I am simplifying all these and similar checks to the style below, which is more in line with what you are suggesting (I think):
`                    p = c.Parameters
                    if p.get("X") and p.get("Y"):
                        # Do Action
`

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 3 above)
Comment (Russ4262) [reply src/Mod/Path/PathTests/TestPathSlot.py:178]: > you're converting a float to a string - could you end up with something like `'(7.499999999999, 0.00000000001)'` ?

I don't think so.  I am rounding values to say three or six decimal places, prior to comparison, to avoid such scenarios.  So long as the `round()` function is working as I am expecting, the subsequent string representation should be only, say, two decimal places max per my rounding argument used.

Does this response address your concern with my methodology, or am I still missing something in the process?

File: src/Mod/Path/PathTests/TestPathSlot.py (hunk 5 above)
Comment (Russ4262) [reply src/Mod/Path/PathTests/TestPathSlot.py:68]: >... maybe use multiple files, it's easier to verify multiple simple shapes than one big one which includes everything. ...

My thinking is to use one file, but it will contain multiple bodies to be used as bases for Jobs in this Slot unit test.  This way I think I could add another body with more targeted slots related to the attached image above without having to alter that particular body. 
.
Does this single-file, multi-body strategy seem okay?
